byterun/bigarray.c

```diff
@@ -439,22 +439,31 @@ static void caml_ba_deserialize_longarray(void * dest, intnat num_elts)
 CAMLexport uintnat caml_ba_deserialize(void * dst)
 {
   struct caml_ba_array * b = dst;
-  int i, elt_size;
-  uintnat num_elts;
+  int i;
+  uintnat num_elts, size;
 
   /* Read back header information */
   b->num_dims = caml_deserialize_uint_4();
+  if (b->num_dims < 0 || b->num_dims > CAML_BA_MAX_NUM_DIMS)
+    caml_deserialize_error("input_value: wrong number of bigarray dimensions");
   b->flags = caml_deserialize_uint_4() | CAML_BA_MANAGED;
   b->proxy = NULL;
   for (i = 0; i < b->num_dims; i++) b->dim[i] = caml_deserialize_uint_4();
-  /* Compute total number of elements */
-  num_elts = caml_ba_num_elts(b);
-  /* Determine element size in bytes */
+  /* Compute total number of elements.  Watch out for overflows (MPR#7765). */
+  num_elts = 1;
+  for (i = 0; i < b->num_dims; i++) {
+    if (caml_umul_overflow(num_elts, b->dim[i], &num_elts))
+      caml_deserialize_error("input_value: size overflow for bigarray");
+  }
```

Comment:
Here you inlined the `caml_ba_num_elts` function and check internal multiplications for overflow. I find it a bit strange:

- if overflows may occur during `caml_ba_num_elts` computation, then I would rather make the function safe in some way (or at least returning `-1` instead of a nonsensical result)
- if, as I would naively expect, we know that the number of elements can never be large enough to provoke an overflow, then we can just can `caml_ba_num_elts` here, and just keep the check for `elt_size * num_elts` below.

Reply:
In my mind, `caml_ba_num_elts` is to be used for "trusted" bigarrays, those that already passed the non-overflow checks on creation.  There is some run-time cost in checking for overflow in `caml_ba_num_elts`.

> if, as I would naively expect, we know that the number of elements can never be large enough to provoke an overflow,

We have no such guarantee when we unmarshal.  The number and values of dimensions may be completely arbitrary.  All multiplications must be guarded against overflow.

Reply:
Oh, that was a silly question, sorry.